### PR TITLE
8313574: GenShen: Completing a global mark should also complete an old mark

### PR DESCRIPTION
**Pull request by earthling-amzn**
The mark context is shared, but each generation maintains a separate flag (`ShenandoahGeneration::_is_marking_complete`) to record the status of the mark. The remembered set scan will use the mark bitmap only if the old generation mark is complete. Otherwise, it walks the old regions and requires them to be parseable.

Presently, we make the heap parseable after global marking finishes, but this work can be deferred to the start of an old mark if we had global mark record completion of old marking.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8313574](https://bugs.openjdk.org/browse/JDK-8313574): GenShen: Completing a global mark should also complete an old mark (**Bug** - P3)

### Reviewers
 * [Kelvin Nilsen](https://openjdk.org/census#kdnilsen) (@kdnilsen - Committer)
 * [Y. Srinivas Ramakrishna](https://openjdk.org/census#ysr) (@ysramakrishna - Committer) ⚠️ Review applies to [2bf50f3f](https://git.openjdk.org/shenandoah/pull/304/files/2bf50f3fe3fc2a25a8679c23f3870a1f13361446)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/shenandoah.git pull/304/head:pull/304` \
`$ git checkout pull/304`

Update a local copy of the PR: \
`$ git checkout pull/304` \
`$ git pull https://git.openjdk.org/shenandoah.git pull/304/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 304`

View PR using the GUI difftool: \
`$ git pr show -t 304`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/shenandoah/pull/304.diff">https://git.openjdk.org/shenandoah/pull/304.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/shenandoah/pull/304#issuecomment-1666013873)

### CODE REVIEW

**Review by kdnilsen — APPROVED** [review]
Looks good.  Thanks for prompt fix.

**Review by ysramakrishna — APPROVED** [review]
LGTM.

Please add a stack retrace of the failure you'd typically see (and perhaps the failing test) that identified this issue into the associated JBS ticket.

Fix looks good.

**Comment by ysramakrishna** [comment]
> LGTM.
> 
> Please add a stack retrace of the failure you'd typically see (and perhaps the failing test) that identified this issue into the associated JBS ticket.
> 
> Fix looks good.

Also I am guessing the JBS issue should be a "bug" rather than just called an enhancement (which I agree it is also :-). Or may be the bug is tracked in a different ticket and will be closed as a duplicate of this enhancement?

**Comment by earthling-amzn** [comment]
Converted to draft because of failures in Dacapo suite.

File: src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.hpp
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.hpp:149]: The return value from `coalesce_and_fill_candidates()` (which by the way could be confused for an imperative action verb) lready available in `last_old_region_index()` although might be a good idea to rename appropriately? Also fine to leave as is, but it sounds to me like one name might work for both uses.

File: src/hotspot/share/gc/shenandoah/shenandoahDegeneratedGC.cpp
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahDegeneratedGC.cpp:105]: Is this not true?

You could rearrange the test into the assertion check though:

```
    assert(!_generation->is_global() || !heap->is_old_gc_active, "Old GC should not be active during global cycle");
```
(or its de-Morganization, if you prefer).

File: src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp
Comment (ysramakrishna) [inline src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp:417]: This transition has arrows at both ends (IDLE and WAITING FOR EVACUATIONS).

**Review by ysramakrishna — COMMENTED** [review]
Left a few comments, and going over all of the changes again to make sure I understand the changes completely again before I sign off on the review.

**Review by ysramakrishna — APPROVED** [review]
Changes LGTM, modulo a few minor comments in previous set of comments.

File: src/hotspot/share/gc/shenandoah/shenandoahDegeneratedGC.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahDegeneratedGC.cpp:105]: With this change, if the global cycle degenerates after final mark there could now be regions waiting to be coalesced and filled, so the old gc state is not idle.

File: src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp:417]: That is intentional, I shall also add an arrow at both ends for IDLE and WAITING FOR FILL. Global cycles do not progress the old generation through marking/bootstrapping so they are allowed to move the old generation from IDLE to one of these states at the end of global mark.

File: src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.hpp
Comment (earthling-amzn) [reply src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.hpp:149]: I agree, one (good) name for both is best.

File: src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp
Comment (ysramakrishna) [reply src/hotspot/share/gc/shenandoah/shenandoahOldGeneration.cpp:417]: I see, so a bidirectional arrow?

It would of course be ideal (in the fullness of time) if there were events labeling the transitions, even if this required a table for the event alphabet.

**Pull request closed without merging**
